### PR TITLE
check ip family for node port connectivity test

### PR DESCRIPTION
**Pull request by aramase**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:
The current `testReachabilityOverNodePorts` tests against all node `InternalIPs`. It's possible a single stack v4/v6 cluster is supported with dual stack hosts in which case the cloud providers would update the `InternalIP` with both. This PR checks the ip family of the `clusterIP` and checks node port connectivity on all `InternalIP` that belong to the same family.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

File: test/e2e/framework/service/jig.go [hunk 1]
```diff
@@ -764,17 +764,31 @@ func testReachabilityOverClusterIP(clusterIP string, sp v1.ServicePort, execPod
 	return nil
 }
 
-func testReachabilityOverNodePorts(nodes *v1.NodeList, sp v1.ServicePort, pod *v1.Pod) error {
+func testReachabilityOverNodePorts(nodes *v1.NodeList, sp v1.ServicePort, pod *v1.Pod, clusterIP string) error {
 	internalAddrs := e2enode.CollectAddresses(nodes, v1.NodeInternalIP)
 	externalAddrs := e2enode.CollectAddresses(nodes, v1.NodeExternalIP)
+	reIPV4, err := regexp.Compile(e2enetwork.RegexIPv4)
```
Comment (neolit123) [inline test/e2e/framework/service/jig.go]: MustCompile seems more appropriate here.

File: test/e2e/framework/service/jig.go [hunk 2]
```diff
@@ -764,17 +764,31 @@ func testReachabilityOverClusterIP(clusterIP string, sp v1.ServicePort, execPod
...
 		if isInvalidOrLocalhostAddress(internalAddr) {
 			framework.Logf("skipping testEndpointReachability() for internal adddress %s", internalAddr)
 			continue
 		}
-		err := testEndpointReachability(internalAddr, sp.NodePort, sp.Protocol, pod)
+		isNodeInternalIPV4 := reIPV4.MatchString(internalAddr)
+		// Check service reachability on the node internalIP which is same family
+		// as clusterIP
+		if isClusterIPV4 != isNodeInternalIPV4 {
+			framework.Logf("skipping testEndpointReachability() for internal adddress %s as does not match cluster ip (%s) family", internalAddr, clusterIP)
```
Comment (neolit123) [inline test/e2e/framework/service/jig.go]: `as does not match` -> `as it does not match`

Comment (neolit123) [reply test/e2e/framework/service/jig.go]: `ip` -> `IP`

**Comment by neolit123** [comment]
@aramase 
thanks for the fix.
please ping someone from #sig-network for review on the logic in this change.
i can approve if they are happy about it.

File: test/e2e/framework/service/jig.go (hunk 1 above)
Comment (aramase) [reply test/e2e/framework/service/jig.go]: updated!

File: test/e2e/framework/service/jig.go (hunk 2 above)
Comment (aramase) [reply test/e2e/framework/service/jig.go]: updated!

File: test/e2e/framework/service/jig.go [hunk 3]
```diff
@@ -764,16 +764,27 @@ func testReachabilityOverClusterIP(clusterIP string, sp v1.ServicePort, execPod
 	return nil
 }
 
-func testReachabilityOverNodePorts(nodes *v1.NodeList, sp v1.ServicePort, pod *v1.Pod) error {
+func testReachabilityOverNodePorts(nodes *v1.NodeList, sp v1.ServicePort, pod *v1.Pod, clusterIP string) error {
 	internalAddrs := e2enode.CollectAddresses(nodes, v1.NodeInternalIP)
 	externalAddrs := e2enode.CollectAddresses(nodes, v1.NodeExternalIP)
+	reIPV4 := regexp.MustCompile(e2enetwork.RegexIPv4)
+	isClusterIPV4 := reIPV4.MatchString(clusterIP)
```
Comment (aojea) [inline test/e2e/framework/service/jig.go]: we have some helpers in utils.net for detecting the ipFamily or we can just parse it with net.ParseIP, I think that is better than the regex for consistency

File: test/e2e/framework/service/jig.go [hunk 4]
```diff
@@ -879,7 +890,7 @@ func (j *TestJig) checkNodePortServiceReachability(svc *v1.Service, pod *v1.Pod)
 		if err != nil {
 			return err
 		}
-		err = testReachabilityOverNodePorts(nodes, servicePort, pod)
+		err = testReachabilityOverNodePorts(nodes, servicePort, pod, clusterIP)
```
Comment (aojea) [inline test/e2e/framework/service/jig.go]: just food for thought, should we pass the clusterIP or a `isIPv6 bool`?

File: test/e2e/framework/service/jig.go (hunk 3 above)
Comment (aramase) [reply test/e2e/framework/service/jig.go]: I was thinking the same thing, then saw regex was used in a couple of other funcs, so used it! But, updated it with `net.ParseIP` now.

File: test/e2e/framework/service/jig.go (hunk 4 above)
Comment (aramase) [reply test/e2e/framework/service/jig.go]: we can do that, but just figured the `testReachabilityOverNodePorts` should make the decision on what it checks and has to do with the `clusterIP`. WDYT?

File: test/e2e/framework/service/jig.go [hunk 5]
```diff
@@ -764,16 +764,26 @@ func testReachabilityOverClusterIP(clusterIP string, sp v1.ServicePort, execPod
...
 			continue
 		}
+		isNodeInternalIPV4 := net.ParseIP(internalAddr).To4() != nil
+		// Check service reachability on the node internalIP which is same family
+		// as clusterIP
+		if isClusterIPV4 != isNodeInternalIPV4 {
+			framework.Logf("skipping testEndpointReachability() for internal adddress %s as it does not match clusterIP (%s) family", internalAddr, clusterIP)
+			continue
+		}
+
```
Comment (aojea) [inline test/e2e/framework/service/jig.go]: lgtm
what about the underneath loop ? 
`	for _, externalAddr := range externalAddrs {`
is ok to left it as is?

File: test/e2e/framework/service/jig.go (hunk 4 above)
Comment (aojea) [reply test/e2e/framework/service/jig.go]: it seems correct one or the other, I don't have a strong opinion about this, this looks ok

File: test/e2e/framework/service/jig.go (hunk 5 above)
Comment (aramase) [reply test/e2e/framework/service/jig.go]: done!

**Pull request merged**